Tabs: unify vertical tabs styles

DaniGuardiola opened the pull request:
Fixes https://github.com/WordPress/gutenberg/issues/65837 (as a side effect)
Fixes https://github.com/WordPress/gutenberg/issues/65863
Fixes https://github.com/WordPress/gutenberg/issues/65865

<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

Unifies the styles of vertical tabs across Gutenberg.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

See https://github.com/WordPress/gutenberg/issues/64307

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

By removing style overrides and tweaking the `Tabs` component to adapt to all needs.

## Follow-ups

- ~Use `transform` https://github.com/WordPress/gutenberg/pull/65387#issuecomment-2384643399~ Already done!
- Uniform tab height https://github.com/WordPress/gutenberg/pull/65387#discussion_r1782494115

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a post or page. -->
<!-- 2. Insert a heading block. -->
<!-- 3. etc. -->

Try the inserter patterns/media vertical tabs.

## Screenshots or screencast <!-- if applicable -->

https://github.com/user-attachments/assets/5845b4b1-4a65-47da-a092-fff3e1fe934c

DaniGuardiola committed:
Remove inserter pattern overrides

DaniGuardiola commented:
@WordPress/gutenberg-design for now I've only removed style overrides. Here's the inserter > patterns `Tabs` instance.

Before (with overrides):

![{A47A17A9-1DF6-4C3A-A208-DE89E363F7BE}](https://github.com/user-attachments/assets/942b2b95-3aac-418c-9a7a-5ed54e61cade)

After (without overrides):

![{93934931-7A34-43D7-99D3-3A09119BA333}](https://github.com/user-attachments/assets/c2425900-9350-434a-9d26-5b2874b4d665)

The main differences I spot in the previous style overrides vs default `Tabs` styles:

- Lighter font weight
- Light blue indicator
- Active tab has text and svg fill color set to blue (vs. no difference)
- Smaller tab padding resulting in shorter/more compact tabs

For reference, here's the preferences dialog tabs, which already don't have any style overrides:

![{1EFFFCA3-461D-4FB5-A33A-50FF46B11D54}](https://github.com/user-attachments/assets/53b40f5f-0e5f-4f38-94bf-c5e90fe9bbb7)

These are the only two instances of Tabs at the moment in Gutenberg. Ideally, we'd find styles that work in both cases. If necessary, we can consider introducing some sort of variant, but let's try to keep that at a minimum and find something that we feel good about regarding future usage of the component, to prevent more style overrides in the future.

Please leave some feedback, happy to implement any ideas so we can test them!

jameskoster commented:
+100 to removing the overrides.

That said, I think the appearance in the Editor is the direction we want to go in. So would it make sense to update the component first, then remove the overrides after? That way there should be minimal visual disruption.

One detail we need to consider is that the color change (gray or blue) is of inadequate contrast to meet requirements, so we need another way to communicate which tab is active. We could address this while updating the component. A good place to start could be making the active tab font weight slightly heavier. 

What do you think?

DaniGuardiola commented:
> So would it make sense to update the component first, then remove the overrides after? That way there should be minimal visual disruption.

Well, the thing is we only have two instances of vertical tabs, so what you're proposing doesn't make much sense purely in terms of the sequence hehe.

Here's how I think this PR should evolve:

1. Remove overrides (done).
2. Update the Tabs component to look how we want it to look in both instances (inserter and preferences).
3. Test and iterate until we feel good about it.
4. Merge and ship it.

I hope that makes sense to you?

Regarding your suggestion that the styles of the inserter tabs are closer to what we want, that makes sense to me. I'll tweak the base component, so both instances will look closer to it.

In parallel, I will also apply your suggestion about the active tab styles. Then, we can take it from there.

Sound good? :)

jameskoster commented:
@DaniGuardiola In my experience we don't tend to update multiple things in a single PR like you're suggesting, hence my idea to update the component first. That said, it does make sense to me in this context, providing other folks in the @WordPress/gutenberg-components team agree.

DaniGuardiola committed:
Merge branch 'trunk' of https://github.com/WordPress/gutenberg into improve/unify-vertical-tabs-styles

DaniGuardiola commented:
> A good place to start could be making the active tab font weight slightly heavier.

Currently, the weight for tabs is 500. In the inserter, it was overridden to 400.

What if we:

- Set the default width to 400 (like in the inserter).
- Make the active tab 500.

This is how that looks like in inserter and preferences:

https://github.com/user-attachments/assets/bee03b24-eefd-4c13-860e-0ee711d1b846

I also tried making the active tab have the theme accent color (blue in this case) like the inserter style overrides did. For reference, this is the same color that tabs have when hovered. Here's how it looks like:

https://github.com/user-attachments/assets/f2235efa-e25b-4635-95a4-9f4f7a8f303a

I quite like this with the light gray indicator. However, it might still not be enough for colorblind or low-vision people.

Here, I tried making the active tab have a weight of 600 (on top of all previous changes):

https://github.com/user-attachments/assets/b69a3ce5-15a4-4b1b-87ac-18b42b3935ac

And here's with 700:

https://github.com/user-attachments/assets/e72cfcc6-ad19-4616-b302-a14d6c9ecf82

I personally feel much more comfortable shipping this with the increased font weights. Thoughts?

DaniGuardiola commented:
By the way, the font weight and text color changes also affect horizontal tabs, as you can see in the videos. That said, we could always special-case it to vertical tabs if we don't want that to happen. Maybe worth doing it in both for consistency though, it communicates more clearly that both (vertical and horizontal) are the same component rather than different ones.

jameskoster commented:
I reckon we should only apply the weight change to vertical tabs, if possible. Horizontal tabs are fine by virtue of the 'active' shape having enough contrast. 

In https://github.com/WordPress/gutenberg/issues/64340#issuecomment-2293705925 we're leaning towards `500` where additional weight is required, so perhaps we try that here, with inactive tabs using `400`?

Here's a visual for the different states, hopefully this communicates things better than words :D 

<img width="774" alt="Screenshot 2024-09-19 at 15 11 49" src="https://github.com/user-attachments/assets/046610f7-bd45-46f6-ac1e-7280cf394206">

[Figma link](https://www.figma.com/design/804HN2REV2iap2ytjRQ055/Core-System-Library?node-id=14192-15206&t=SEplYOrx6xkt52rh-1).

DaniGuardiola commented:
@jameskoster 

> In https://github.com/WordPress/gutenberg/issues/64340#issuecomment-2293705925 we're leaning towards `500` where additional weight is required, so perhaps we try that here, with inactive tabs using `400`?

You can see this in my previous comment, it's the first example. Qs:

1. From the screenshot of the Figma design, I infer that we also want the text color to be the theme accent color as I suggested in my last comment, correct?
2. And we'll want the light blue (theme accent color at 4% opacity) indicator color too, instead of the current gray one?
3. I will insist on the point that 500 is not enough weight in my opinion to indicate the active tab. The active text color (as mentioned in point 1 of this list) does help a bit, but it doesn't seem like enough in this context. I got good vision, and still struggle a bit to clearly tell those apart. See my last comment for examples at higher weights.

For now, I'm gonna assume that the answer is yes to points 1 and 2 and push those changes in this PR.

FYI, there's a problem that https://github.com/WordPress/gutenberg/pull/64371 fixes that makes the indicator not show up in the inserter, in case you wanted to try it locally. I'm performing a quick hack to get these screenshots and videos for now, but you won't be able to test it locally for now. Hopefully we'll be able to merge that soon though. Let me know if you want me to record or capture something in the meantime.

DaniGuardiola committed:
Make font weights 400 (inactive) and 500 (active)

DaniGuardiola commented:
@jameskoster 

I applied all changes, including the 400/500 weight change, and here's how that looks like currently. 500 (semibold) still seems too low for me, especially with the light blue indicator which looks even dimmer than the previous gray one to me.


https://github.com/user-attachments/assets/a54384df-27f4-4fa0-a77c-9fcb30b86c50

DaniGuardiola committed:
Apply styles only when vertical.

DaniGuardiola committed:
Make vertical indicator theme accent color at 4% opacity.

jameskoster commented:
You may be right about the weight, but the more we increase it the more legibility in certain languages decreases. I think we should proceed with `500` for now, as it is aligned with https://github.com/WordPress/gutenberg/issues/64340#issuecomment-2293705925 and also because it is objectively an improvement on trunk.

We can gather feedback and open a dedicated issue for this detail. There are other options to explore like adding a shape to the active tab:

<img width="262" alt="Screenshot 2024-09-20 at 12 29 47" src="https://github.com/user-attachments/assets/09e3a927-9f41-46c3-babf-5ba4ac997dae">

It's not something that needs to block the other enhancements here, and needs broader visibility to determine the best direction.

jameskoster commented:
One small detail; the radius on the active tab should be `radius-small`.

DaniGuardiola commented:
@jameskoster various thoughts:

- Font weight: alright, sounds good.
- Radius: we could add it, but FYI due to constraints of the web platform, adding a radius would force us to downgrade to a different animation technique for the indicator that would result in slightly worse performance. Not the end of the world, but important to consider.
- The proposed border design has a curvature that, as far as I know, is not doable in CSS. The best that can be done (without some ugly convoluted tricks that would probably have a negative effect on perf and bundle size) is something like the image below. I strongly recommend that we don't try to implement that non-standard border curvature.

<img width="944" alt="image" src="https://github.com/user-attachments/assets/6acdf81f-a209-4257-a478-61910171d911">

- Current focus-visible style looks like the image below. This is the same style as in horizontal tabs. Is this okay? I personally think it's fine and better since it's consistent with the other orientation. Also less divergent code (for horizontal vs vertical).

<img width="892" alt="image" src="https://github.com/user-attachments/assets/a70114d9-e258-44cf-a06c-a85ea6d9f108">

- Are we okay with the size as it currently is? The most obvious change in the patterns inserter after the merges would currently be that those tabs are significantly taller now and less dense. I think it's fine, personally, but something to discuss anyway.

jameskoster commented:
@DaniGuardiola sorry, the border treatment wasn't a proposal; just an example of something we could try instead of changing the font weight. It's best to explore this in detail in a dedicated issue.

> slightly worse performance

How bad are we talking? The majority of UI components have rounded corners, it should be a shame to introduce an inconsistency here.

jameskoster commented:
> Sure, though I'm a bit worried that we'll make tabs worse if we merge this PR with a design that's considered "incomplete".

I don't think so because we're objectively improving compared with trunk in terms of a11y.

> This is the same style as in horizontal tabs. Is this okay?

Agree with you that this is okay for now.

> Are we okay with the size as it currently is?

Ideally the vertical tabs are 40px tall. This aligns with the general size convention we have for interactive elements (24/32/40).

DaniGuardiola commented:
> I don't think so because we're objectively improving compared with trunk in terms of a11y.

I find the light blue background harder to see than the previous gray one. This only affects the preferences tabs FWIW. That said, no strong feelings on my side, as long as we don't forget to improve a11y in a follow-up.

> Agree with you that this is okay for now.

Cool!

> Ideally the vertical tabs are 40px tall.

Okay, right now they are 48px tall. I'll change this to 40px. For reference, this is the original size of the patterns/media tabs, ergo after this PR merges, the size change will only be noticeable on the preferences tabs.

DaniGuardiola committed:
Make height 48px.

DaniGuardiola committed:
Add radius.

DaniGuardiola committed:
Also use hover styles in focus-visible.

DaniGuardiola commented:
@jameskoster 

I've implemented the following changes:

- Size is now 40px.
- Indicator has "radius-small" radius now.
- It's a small detail, but now focus-visible states have a similar style to hover states, specifically the text color. This is usually a good idea as it brings consistency to states that essentially mean the exact same thing in different mediums (pointer & keyboard).

Here's the current state:

https://github.com/user-attachments/assets/1712616c-d49a-4ce7-81a4-4f1834771a2a

Is there anything else you feel like we need to tweak or add before considering a merge? cc @WordPress/gutenberg-design @WordPress/gutenberg-components

jameskoster commented:
The editor preferences modal is fine, but I'm not seeing the background on active tabs in the Inserter:

<img width="346" alt="Screenshot 2024-09-25 at 09 48 30" src="https://github.com/user-attachments/assets/1d7cc9a4-0669-4a6e-aaa4-31e5201e7d1a">

Horizontal padding is a bit too big; let's make that `12px`. Ideally `padding-right` reduces to `4px` when the suffix (`>` icon) is present. This will maintain alignment with trunk.

Sorry to flip-flop on this, but do you think we could make the focus ring encircle the full tab footprint, like on trunk? IE:

<img width="282" alt="Screenshot 2024-09-25 at 09 53 09" src="https://github.com/user-attachments/assets/b554e047-8cc9-4096-9d91-5d1d2d61d6ed">

> Size is now 40px.

Thank you, there's a small caveat here—we should probably use min-height so that in rare cases where the label is very long the tab can grow to accommodate. Currently there can be awkward overlaps:

<img width="362" alt="Screenshot 2024-09-25 at 09 57 11" src="https://github.com/user-attachments/assets/24503097-13eb-4a0a-a21e-6549f6830a0d">

DaniGuardiola commented:
@jameskoster 

> I'm not seeing the background on active tabs in the Inserter

Yes, that's expected. See what I commented [earlier in the thread](https://github.com/WordPress/gutenberg/pull/65387#issuecomment-2361573309):

"FYI, there's a problem that https://github.com/WordPress/gutenberg/pull/64371 fixes that makes the indicator not show up in the inserter, in case you wanted to try it locally. I'm performing a quick hack to get these screenshots and videos for now, but you won't be able to test it locally for now. Hopefully we'll be able to merge that soon though. Let me know if you want me to record or capture something in the meantime."

To clarify, this is not a bug currently affecting us, as it's only visible when you undo the style overrides of the inserter (which we do in this PR). FWIW, I'm considering moving applying those changes here anyway, because that PR is taking longer to figure out and merge than expected. That'd allow you to test locally, though as I said, I'm happy to take any screen captures and videos necessary in the meantime for design feedback, just let me know.

> Horizontal padding is a bit too big; let's make that 12px

Sure.

> Ideally padding-right reduces to 4px when the suffix (> icon) is present. This will maintain alignment with trunk.

Well, the icon is currently not part of the Tabs component itself, it's just part of its children. There's no trivial way to achieve this with the current setup, however we could consider adding support for this icon in the component itself, maybe as a prop of `Tabs.Tab` (e.g. `withArrow`), or as a more generic `icon` prop if we think it's worth it. If we do so, the dynamic padding is easier to implement and less messy. What do you think?

> Thank you, there's a small caveat here—we should probably use min-height so that in rare cases where the label is very long the tab can grow to accommodate. Currently there can be awkward overlaps

True. I guess I made the height fixed for consistency with horizontal tabs, but in this case I guess it doesn't make much sense to let it extend horizontally like we do in horizontal orientation. I'll make it a min-height.

DaniGuardiola committed:
Fix indicator not visible in inserter > patterns/media.

DaniGuardiola committed:
Adjust padding.

DaniGuardiola committed:
Tweak focus ring.

DaniGuardiola committed:
Wrap long labels.

DaniGuardiola commented:
Applied all of your feedback, and ported the patterns tabs indicator fix here so you can test locally right away. Let me know if there are any other tweaks you want to try :) @jameskoster

DaniGuardiola commented:
Since this looks close now, I'm adding the team for code reviews and undrafting the PR. I will add a changelog before merge!

ciampo commented:
> > Ideally padding-right reduces to 4px when the suffix (> icon) is present. This will maintain alignment with trunk.
> 
> Well, the icon is currently not part of the Tabs component itself, it's just part of its children. There's no trivial way to achieve this with the current setup, however we could consider adding support for this icon in the component itself, maybe as a prop of `Tabs.Tab` (e.g. `withArrow`), or as a more generic `icon` prop if we think it's worth it. If we do so, the dynamic padding is easier to implement and less messy. What do you think?

I personally like to keep separation of concerns — meaning that `Tabs` should only know about tab-specific things, and otherwise offer easy ways for its consumers to extend it.

The risk of adding an `icon` prop is that folks may start to want flexibility on where to render it, and so we could end up in a situation like `Button`, where we have an `icon` and `iconPosition` prop, and folks still sometimes just render a custom icon as part of `children`.

I have a couple of ideas:

### Adding a `suffix` prop

Maybe we could strive a good balance by adding a `suffix` prop:
- when such a prop is defined, the corresponding padding is not applied to the tab (maybe we could still make it available via a CSS prop, in case the consumer wants to apply that same padding);
- it allows consumers to extend `Tabs.Tab`, but at the same time it doesn't add a very specific prop (`icon`) and a ad-hoc layout rule (the different padding)

The consumer of `Tabs.Tab` could then also make their own reusable version which always shows a chevron and can be reused across their app.

### Adding a `showChevron` (name TBD) prop

This is a more specific alternative, but it still feels "generic" enough to `Tabs` functionality. We'd need to understand if that applies at all to the horizontal orientation too.


---

Also, is it worth waiting for #64371 to be merged, before testing this PR?

jameskoster commented:
A `suffix` prop might work, but it's tricky because the adjusted padding is only necessary when the suffix is an icon, and the chevron icon specifically.

Thinking about this a bit more... the chevron is mostly used as an indicator for the behavior... Since this PR is about unifying tabs; if we feel the chevrons are necessary in the pattern category tabs then shouldn't they be applied to _all_ instances of vertical tabs? It would be good to be consistent about this. 

I'm wondering if there could be middle-ground where we show them on hover and when the tab is active:

https://github.com/user-attachments/assets/7d0406d0-0707-4a02-abf6-810dc7ecda8a

I'd welcome thoughts from @WordPress/gutenberg-design on that.

@DaniGuardiola if we applied this at the component level then we wouldn't need to adjust the font-weight for the active tab; the shape change (visible chevron) should be adequate in meeting a11y requirements.

jasmussen commented:
> I'm wondering if there could be middle-ground where we show them on hover and when the tab is active:

I think this could be good, the chevron would also be an active/focus shape change. Further, I would think we should make the overall list of categories less wide, so the aggregate width of categories + flyout tray is not quite as huge as it is today (not in this PR!)

jameskoster commented:
Yes the chevron seems to be permanently visible partly to add some visual balance. Once we make the panel narrower that shouldn't be necessary.

ciampo commented:
Let's make sure that such a chevron could apply in _all_ cases. For example, what if the consumer wants to display tabs to the right-hand side of the tabpanels?

Just trying to play devil's advocate

ciampo commented:
We could also add a chevron-related prop accepting `inline-start`, `inline-end`, `block-start` and `block-end` as possible values. The values would indicate the direction that the chevron points to. When the prop is `undefined`, no chevron is shown. We could adopt `inline-end` (ie. right, for LTR languages) as the default value for `horizontal` orientation, and `undefined` for vertical orientations.

That would allow consumers to customize the direction of the chevron should they need to, or to remove it altogether.

Could this represent a good balance?

DaniGuardiola commented:
I like @jameskoster's suggestion, it's the simplest one since it reduces potential for divergence. The RTL question @ciampo poses is relevant, however I believe it's not something we necessarily need to worry about _right now_ since we do not have a use-case for RTL tabs currently.

I do think we need to make sure whatever we implement now doesn't get in the way of RTL support in the future. A prop like you're proposing seems implementable at a later time (I would also consider using `:dir()` so we have one less prop).

What do you think? For now, I'm gonna implement @jameskoster's suggestion to see how it feels, while we discuss the RTL thing.

DaniGuardiola commented:
Huh, actually I just realized we currently do support RTL in the inserter > patterns tabs.

<img width="648" alt="image" src="https://github.com/user-attachments/assets/fc0d089e-b74b-4e25-bbac-4fc56be8b728">

I think we can get away with using `:dir()` though, so let me try that.

DaniGuardiola committed:
Add chevron and fix a few minor details.

DaniGuardiola commented:
Alright, I've implemented the changes. This includes removing the increased font-weight on the selected tab.

I also fixed a couple of things I noticed during my testing. RTL is implemented through `:dir()` and seems to work great for RTL locales like arabic in the inserter > patterns/media tabs. Let me know what you think and if you find any issues.

DaniGuardiola commented:
Btw, I noticed this behavior where the inserter and tabs jump around on initial selection:

https://github.com/user-attachments/assets/b275910c-3d1f-458c-8549-9b1ebd16b702

This is trunk, so it's unrelated to this PR. Will likely report in a separate issue.

jameskoster commented:
A prop for the chevron feels like something to explore separately to me. It could work, but to meet a11y requirements we'll need another shape change on the active state as a fallback for consumers who remove the chevron.

This seems very close to me, I noticed one issue: When I open the Inserter the 'All' pattern category is showing an unexpected mix of selected and unselected... there's no blue accents, but I can see the chevron:

<img width="349" alt="Screenshot 2024-09-30 at 10 30 05" src="https://github.com/user-attachments/assets/75195630-c06c-40f6-a9e2-b5fb3b4dcea3">

DaniGuardiola committed:
Merge branch 'trunk' of https://github.com/WordPress/gutenberg into improve/unify-vertical-tabs-styles

DaniGuardiola committed:
Fix merge issues.

DaniGuardiola committed:
Fix focus indicator (gets cropped with the new overflow auto setting)

DaniGuardiola committed:
Merge branch 'trunk' of https://github.com/WordPress/gutenberg into improve/unify-vertical-tabs-styles

DaniGuardiola commented:
@jameskoster good catch, this was caused by a quirk of the implementation, switched to a different strategy in styling that seems to work.

Re: the prop, agree that additional configuration like removing the chevron should be explored separately. My initial feelings about that is that I'm not sure it's a good idea. The chevron looks fine to me and I don't think it'd be necessary to make it customizable.

DaniGuardiola committed:
Fix unwanted chevron.

DaniGuardiola commented:
Btw, I plan on following this up with a PR to switch back to transform for the animation, since we figured out a way to make it work with border-radius, so the performance implication I discussed in https://github.com/WordPress/gutenberg/pull/65387#issuecomment-2365363391 will be mitigated very soon.

ciampo commented on packages/components/src/tabs/styles.ts:
```diff
@@ -181,10 +174,50 @@ export const Tab = styled( Ariakit.Tab )`
 
-	[aria-orientation='vertical'] & {
+	[aria-orientation='horizontal'] & {
+		padding: ${ space( 4 ) };
 		min-height: ${ space(
-			10
+			12
 		) }; // Avoid fixed height to allow for long strings that go in multiple lines.
+		height: auto;
```
Also in the context of https://github.com/WordPress/gutenberg/issues/65624, the changes from this PR make horizontal tabs with Icons grow from `56px` to `59.5px` tall.

![Screenshot 2024-10-01 at 12 03 20](https://github.com/user-attachments/assets/191f9829-f917-455c-9546-b920829e4b49)

I'm ok with tacking this separately, but let's not forget about it

DaniGuardiola replied:
This is fixed now in this PR.

ciampo commented:
Also, are we ok with this behavior when the tab text doesn't fit in one line?

![Screenshot 2024-10-01 at 12 06 14](https://github.com/user-attachments/assets/e5874f55-bb31-4bf2-bbf6-80f0ef54bd71)

jameskoster commented:
> Also, are we ok with this behavior when the tab text doesn't fit in one line?

Seems okay for now. We can always revisit. I think guidelines should discourage long tab labels generally.

DaniGuardiola commented:
> While playing with the Storybook example, I noticed that moving tab selection (both with keyboard and pointer) can sometimes, very briefly, cause the scrollbar to flash:

This only happens with long overflowing labels, right?

ciampo commented:
> This only happens with long overflowing labels, right?

When the label doesn't fit on one line. I haven't checked it it would happen with any arbitrary content that is taller that the default line height

DaniGuardiola commented:
@jameskoster are we good on the design side to merge?

@ciampo I'm gonna investigate the issue you found, however since the wrapping behavior is more of a failsafe than anything else I think it should be fine to merge given that it's a minor visual glitch, what do you think?

ciampo commented:
> @ciampo I'm gonna investigate the issue you found, however since the wrapping behavior is more of a failsafe than anything else I think it should be fine to merge given that it's a minor visual glitch, what do you think?

Sure, let's treat it as non-blocking, but follow-up to make sure we don't leave it behind

DaniGuardiola commented:
@ciampo sgtm, can I get an approval? I'll merge once/if @jameskoster confirms we're okay design-wise.

ciampo commented:
There seems to be something wrong with the indicator in the patterns inserter — all of the CSS variables are `0` and the indicator is, therefore, not visible:

https://github.com/user-attachments/assets/8a5f7d7f-d26c-45f9-a4fe-d6f623ead41c


It works fine in Storybook and in the preferences modal:

![Screenshot 2024-10-02 at 12 18 26](https://github.com/user-attachments/assets/c2ba1bf4-1812-4087-bda5-5c189483a6f1)

jameskoster commented:
I'm not seeing the issue Marco mentioned above 🤔 So far as I can see this is good to go. We might consider swapping to the small chevron, but that's something we should look into separately.

stokesman commented:
> There seems to be something wrong with the indicator in the patterns inserter

I’m seeing the same and the nested overflow of #65837. Here’s a screen recording of Firefox but I see the same in Chrome.

https://github.com/user-attachments/assets/b5c38a1d-307e-495f-88ad-5be58e7f3b27

DaniGuardiola commented:
I managed to repro. I see how this slipped by all of us - if you don't make the window short enough you don't notice the effect. I will fix it in this PR since it's already quite close. Thanks again for bringing this to my attention @stokesman.

DaniGuardiola commented:
@ciampo I can't repro your bug report. Can you ensure you're in latest with clean env and retry? I'm gonna merge trunk here too.

DaniGuardiola committed:
Merge branch 'trunk' of https://github.com/WordPress/gutenberg into improve/unify-vertical-tabs-styles

DaniGuardiola commented:
Btw, do y'all think this deserves a changelog entry? I think I will add one for the bug(fix) that @stokesman reported, but other than that I'm not sure.

DaniGuardiola committed:
Fix unwanted nested scrollbar in inserter > patterns/media vertical tabs.

DaniGuardiola commented:
@stokesman the bug is fixed now, please test :)

DaniGuardiola committed:
Switch to transform for performance.

DaniGuardiola committed:
Adjust border-radius based on scaling factor.

DaniGuardiola commented:
@ciampo this bug you reported has been fixed now: https://github.com/WordPress/gutenberg/pull/65387#issuecomment-2385367820

The fix involved switching to transform (with the border-radius adjustment too), so that's one thing less to follow-up on.

The only thing remaining is this bug you reported that I (nor @jameskoster) can repro: https://github.com/WordPress/gutenberg/pull/65387#issuecomment-2388292226

stokesman commented on packages/block-editor/CHANGELOG.md:
```diff
@@ -2,6 +2,10 @@
 
 ## Unreleased
 
+## Bug Fixes
+
+-   Fixed a regression that caused the vertical tabs in the inserter (inside the patterns and media sections) to have a nested scrollbar ([#65387](https://github.com/WordPress/gutenberg/pull/65387)).
+
```
Well #64371 was only merged two days ago and there’s been no release since right? My apologies if labeling #65837 "regression" was wrong. It would have to actually be released before that label should be applied?

DaniGuardiola replied:
Alright fair, I will remove the entry 👍

stokesman commented on packages/block-editor/src/components/inserter/style.scss:
```diff
@@ -214,8 +214,9 @@ $block-inserter-tabs-height: 44px;
 .block-editor-inserter__media-tabs-container,
 .block-editor-inserter__block-patterns-tabs-container {
+	flex-shrink: 0;
 	padding: $grid-unit-20;
-	height: 100%;
+	min-height: 100%;
```
Thanks for fixing this! From what I tested `flex: 1` suffices and neither `flex-shrink: 0` nor `min-height` is needed here. If I’m wrong about `flex-shrink`, I’d still prefer `flex: 1 0`.

DaniGuardiola replied:
Sure, let's do that 👍

DaniGuardiola committed:
Apply feedback.

ciampo commented on packages/components/src/tabs/styles.ts:
```diff
@@ -181,10 +185,50 @@ export const Tab = styled( Ariakit.Tab )`
+
+	[aria-orientation='vertical'] & {
+		padding: ${ space( 2 ) } ${ space( 3 ) };
+		padding-left: ${ space( 3 ) };
+		padding-right: ${ space( 3 ) };
```
Should we use logical props here instead?

DaniGuardiola replied:
Makes no difference. I wish the `padding-x` or `padding-inline` shorthand property existed though 🤷

ciampo replied:
> Makes no difference.

True in this case. I just think it's good practice to think logical-first when the option is available.

> I wish the padding-x or padding-inline shorthand property existed though 🤷

It exists: [`padding-inline`](https://developer.mozilla.org/en-US/docs/Web/CSS/padding-inline)

DaniGuardiola replied:
LMAO!

Edit: done.

ciampo reviewed (approved):
> The only thing remaining is this other bug you reported that I (and @jameskoster) can't repro: [#65387 (comment)](https://github.com/WordPress/gutenberg/pull/65387#issuecomment-2388292226)

Maybe the most recent commits have fixed it, or maybe I was hallucinating 🤷  In any case, I can't repro the bug anymore on the latest commit. And also the "glitchy" scrollbar bug posted above is gone too 🎉 

> Btw, do y'all think this deserves a changelog entry? I think I will add one for the bug(fix) that @stokesman reported, but other than that I'm not sure.

- No need for a "bug fix" entry, since that never really got shipped to a version of gutenberg
- I would still add an entry for the change to the tabs styles when in vertical orientation

In any case, none of this feedback is blocking.

LGTM 🚀

DaniGuardiola committed:
Add changelog entry.

jameskoster commented:
Thanks for all the work on this one @DaniGuardiola :)

DaniGuardiola committed:
Merge branch 'trunk' of https://github.com/WordPress/gutenberg into improve/unify-vertical-tabs-styles

DaniGuardiola committed:
Switch to `padding-inline`.

DaniGuardiola committed:
Merge branch 'trunk' of https://github.com/WordPress/gutenberg into improve/unify-vertical-tabs-styles

stokesman commented:
19.4.0 RC1 went out this morning so I added the label to backport this. Correct me if that’s mistaken 🙇.

stokesman commented on packages/components/src/tabs/styles.ts:
```diff
@@ -199,8 +199,7 @@ export const Tab = styled( Ariakit.Tab )`
 		padding: ${ space( 2 ) } ${ space( 3 ) };
-		padding-left: ${ space( 3 ) };
-		padding-right: ${ space( 3 ) };
+		padding-inline: ${ space( 3 ) };
```
Not a big deal and maybe I'm overlooking something but it’s not clear why `padding` shorthand is used first and then `padding-inline`. The latter wouldn’t seem to change anything. If logical properties are preferred then this would be more clear:
```ts
		padding-block: ${ space( 2 ) }
		padding-inline: ${ space( 3 ) }
```
But it seems like in this case the first `padding` alone suffices.

DaniGuardiola replied:
good catch, this was a quick commit but you're right it doesn't make sense, i'll fix it since there's a weird test error i need to fix before merge anyway 👍

DaniGuardiola replied:
Removed the unnecessary property 👍

t-hamano commented:
Regarding font weight, it will vary depending on the OS and default font of the browser. Therefore, for reference, I have attached how it looks on Windows OS.

## Chrome

The font weight becomes thinner.

<details><summary>Details</summary>

### Before

![chrome-trunk](https://github.com/user-attachments/assets/a27e845c-b4b4-4dc7-a20c-a3cd5a6b0e71)

### After

![chrome](https://github.com/user-attachments/assets/75385ed3-2c64-4f8e-8051-803fc39c3985)

</details> 

## Firefox

The font weight remains the same.

<details><summary>Details</summary>

### Before

![firefox-trunk](https://github.com/user-attachments/assets/05b5b7ce-5d61-43f0-a010-3f68e29a9fa4)

### After

![firefox](https://github.com/user-attachments/assets/7f99f9ec-097e-44b3-9df4-7af1407e0dbe)

</details> 

## Chrome (CJK Language)

The font weight becomes thinner.

<details><summary>Details</summary>

### Before

![chrome-trunk-ja](https://github.com/user-attachments/assets/b02d7e2b-bfa3-4c20-a077-cc987210a561)

### After

![image](https://github.com/user-attachments/assets/fa4f848e-68f7-4bc7-90aa-9e1857253ae3)


</details>

t-hamano commented:
Another thing I noticed is that the padding in the Block Inspector tab has changed:

| trunk | This PR |
|--------|--------|
| ![image](https://github.com/user-attachments/assets/0f5a1c0b-a2b7-48fb-add8-c61a7fb990c1) | ![image](https://github.com/user-attachments/assets/84baf8af-7644-48b6-9f4c-a67ff7d1e3a9) | 

We might want to rebase to see if this PR is the problem. I tried rebasing locally and it seems to conflict with trunk.

DaniGuardiola commented:
Thank you! Btw, if someone knows why the test is failing, that'd be helpful. @jeryj and I paired on it and realized it's the new children layout that causes this, apparently only fails in safari. Very odd focus behavior. If someone has a clue that'd be helpful as it's blocking merge - in any case I'll be debugging later today so we can hopefully merge soon.

ciampo commented:
> Another thing I noticed is that the padding in the Block Inspector tab has changed:
> 
> We might want to rebase to see if this PR is the problem. I tried rebasing locally and it seems to conflict with trunk.

Thank you for flagging! That's something we already took note of. We plan to follow-up on it shortly after merging this PR: https://github.com/WordPress/gutenberg/pull/65387#discussion_r1782494115

ciampo commented:
> it's the new children layout that causes this, apparently only fails in safari. Very odd focus behavior. If someone has a clue that'd be helpful as it's blocking merge - in any case I'll be debugging later today so we can hopefully merge soon.

@DaniGuardiola that's weird. I'd try to replicate on `trunk`, because I'd expect such a change to have been introduced in https://github.com/WordPress/gutenberg/pull/65385 (with the `tabindex` attribute on the tablist and the focus management ). I'd also try to understand if it happens with `selectOnMove={false}`, maybe it gives us more info.

DaniGuardiola committed:
Remove unnecessary styles.

DaniGuardiola committed:
Fix horizontal tabs height.

DaniGuardiola committed:
Remove more unnecessary styles (padding).

DaniGuardiola commented:
@t-hamano @ciampo I fixed the padding and height issue. This PR now should not change it for the horizontal tabs.

DaniGuardiola committed:
Make horizontal padding specific to inline.

DaniGuardiola committed:
Make flex/whitespace styles more explicit.

DaniGuardiola committed:
Make scroll margin specific to vertical tabs.

DaniGuardiola committed:
The "inline" in inline-flex is unnecessary and confusing, removed it.

DaniGuardiola committed:
Remove unnecessary position: relative

DaniGuardiola committed:
Make resets more explicit

DaniGuardiola committed:
Remove unnecessary text-align.

DaniGuardiola committed:
Improve comment

DaniGuardiola committed:
Remove unnecessary margin-left

DaniGuardiola committed:
Clean up TabList styles.

DaniGuardiola committed:
Adjust text-align.

DaniGuardiola committed:
Clean up selector

DaniGuardiola committed:
Fix focus indicator

DaniGuardiola committed:
Clean up position: relative.

DaniGuardiola committed:
Fix typo.

DaniGuardiola committed:
Add position: relative back.

DaniGuardiola commented:
I tweaked the focus indicator in vertical tabs to address https://github.com/WordPress/gutenberg/issues/65865 cc @jasmussen @jeryj

Now it animates with the selection indicator when `selectOnMove` is enabled.

What do you think?


https://github.com/user-attachments/assets/ab8a66ed-9736-4778-a976-60a88393391a

DaniGuardiola committed:
Improve focus indicator when selectOnMove is enabled.

DaniGuardiola commented:
Here's a video taken in wp directly


https://github.com/user-attachments/assets/0e2e2bfd-f851-42ab-95cd-2b3fba8f0992

DaniGuardiola committed:
Merge branch 'trunk' of https://github.com/WordPress/gutenberg into improve/unify-vertical-tabs-styles

DaniGuardiola committed:
Add fade in effect to chevron when selectOnMove is enabled.

DaniGuardiola committed:
Use [data-focus-visible] consistently.

DaniGuardiola committed:
Styles clean up.

DaniGuardiola committed:
Add comment for clarity.

DaniGuardiola committed:
Move scroll-margin to the right place.

DaniGuardiola committed:
Use CSS variable for accuracy.

DaniGuardiola committed:
Fix overflow.

ciampo commented:
Gave this a quick round of testing, and noticed some flashing styles for the focus ring in vertical orientation, when select on move is `true`.

Apart from that, I couldn't spot other issues — although we'd definitely need to test thoroughly, as the devil's in the details.

Finally, re. the e2e tests failures, it seems like it is a Safari quirk, and that the modal focus management is broken anyway. I'd be in favour of disabling that specific e2e test on Safari.

DaniGuardiola committed:
Skip failing test for Safari :(

DaniGuardiola committed:
Fix flashing issue.

DaniGuardiola committed:
Transition chevron only on selected and not on hover or focus-visible.

stokesman commented on packages/components/src/tabs/styles.ts:
```diff
@@ -251,11 +251,7 @@ export const TabChevron = styled( Icon )`
-			)
+			[role='tab']:is( [aria-selected='true'],  )
 			& {
 			transition: opacity 0.3s ease-in;
```
This works well enough so feel free to disregard this comment. It’s super subtle but I think a small delay is better. Mechanically the delay should match the transition duration of the indicator (`0.2s` I think). However, to make sure it doesn’t feel like the icon appearance is late the delay can be less than that. Here’s my suggestion:
```css
transition: opacity 0.15s 0.15s linear;
```
I also recommend `linear` easing because on a quick `opacity` transition `ease-in` is imperceptible (though without the delay it might helpful in this case).

DaniGuardiola replied:
We actually tried a few things including delay and different easings, and the current one felt decent. However if you've tested this value live and feel like it's an improvement, I'm happy to change it, since we don't feel too strongly about it anyway :)

Let me know! Really appreciate your input 👍

stokesman replied:
Thanks! I did test but won’t argue it’s a significant improvement. It works as is. Non-linear easing on opacity or color transitions always invites my skepticism but in this case `ease-in` is a pseudo delay.

DaniGuardiola replied:
Forgot to say, but I implemented your transition parameters. Thank you again!

stokesman commented:
I'm not seeing any issues in testing. I think I did see the flashing focus ring Marco mentioned but not now (since ece5efc1d24d209f57f0b235a372b1ef586361cc).

DaniGuardiola commented:
@stokesman yes, the flashing has indeed been fixed 👍 

FYI, https://github.com/WordPress/gutenberg/pull/65878 should be backported before this one to avoid nasty conflicts :)

ciampo commented on packages/components/src/tabs/styles.ts:
```diff
@@ -104,65 +102,87 @@ export const TabListWrapper = styled.div`
-		background-color: ${ COLORS.theme.gray[ 100 ] };
+		}
+		&[data-select-on-move='true']:has(
+				:is( :focus-visible, [data-focus-visible] )
```
This is probably fixed by https://github.com/ariakit/ariakit/pull/4094 which is included in `@ariakit/react` `0.4.11`.

We should follow-up to update ariakit-related dependencies, and consequently we should be able to remove `:focus-visible` 🤞

ciampo reviewed (approved):
Code changes LGTM, and test well in Storybook and in the editor 🚀 

We'll just need to move the CHANGELOG entry under the `Unreleased` section.

ndiego commented:
Hey folks, this is labeled as an enhancement, but it fixes a [regression](https://github.com/WordPress/gutenberg/issues/65837) in 6.7. Therefore, it's unclear to me if this should be included in 6.7 itself. Thoughts?

ciampo commented:
> Hey folks, this is labeled as an enhancement, but it fixes a [regression](https://github.com/WordPress/gutenberg/issues/65837) in 6.7. Therefore, it's unclear to me if this should be included in 6.7 itself. Thoughts?

I thought the regression was introduced in https://github.com/WordPress/gutenberg/pull/64371, which (if I understand correctly ) is not part of 6.7 either — that's why this is marked only to be backported to Gutenberg RC. But if the fix is needed in 6.7, we should definitely backport it (together with #65837 and any dependant PR)

ndiego commented:
> I thought the regression was introduced in https://github.com/WordPress/gutenberg/pull/64371, which I thought was not part of 6.7 either — that's why this is marked only to be backported to Gutenberg RC. But if the fix is needed in 6.7, we should definitely backport it (together with https://github.com/WordPress/gutenberg/issues/65837 and any dependant PR)

You are correct, I just confirmed. I'll remove https://github.com/WordPress/gutenberg/issues/65837 from the 6.7 board. Pardon the false alarm.

DaniGuardiola committed:
Improve chevron opacity transition with suggested value.

DaniGuardiola committed:
fix changelog

DaniGuardiola committed:
Merge branch 'trunk' of https://github.com/WordPress/gutenberg into improve/unify-vertical-tabs-styles

jameskoster commented:
Thanks for all the work on this one @DaniGuardiola 🥳

cbravobernal commented:
I just cherry-picked this PR to the release/19.4 branch to get it included in the next release: 95862263e7

t-hamano commented on packages/components/src/tabs/styles.ts:
```diff
@@ -175,23 +195,69 @@ export const Tab = styled( Ariakit.Tab )`
 
 			@media not ( prefers-reduced-motion ) {
-				transition: opacity 0.1s linear;
+				transition: opacity 0.15s 0.15s linear;
```
Just to be sure, is this intentional? This means the duration is 0.15s and the delay is 0.15s. The `Button` component has a duration of 0.1s and no delay, which feels a bit unnatural to me:

https://github.com/user-attachments/assets/dd4f935a-dda9-40eb-8291-9ba5f20da819

ciampo replied:
It is: https://github.com/WordPress/gutenberg/pull/65387#discussion_r1788055751

I believe the idea is to show the chevron only after the "indicator" has animated to the new selected tab

t-hamano replied:
I saw the comment, but isn't that animation only expected for vertical tabs?

Additionally, I don't know if it's related to this PR, but I can see that the pattern categories in the block sidebar don't seem to animate properly - they don't have a background color when focused:


https://github.com/user-attachments/assets/f2ba1bc3-06ad-4bb2-89b9-926ca03f4235

ciampo replied:
Thank you for reporting - I expect @DaniGuardiola to take a look at this when he's back (I don't think he's online today)

stokesman replied:
> but isn't that animation only expected for vertical tabs?

This seem like a good point. The delay (and maybe the whole transition) should only be needed for `data-select-on-move='true'` conditions, I think.

DaniGuardiola replied:
On my phone, and on sick leave, but just wanted to say that as far as I remember, this transition was only applied to the Chevron in vertical tabs, and only when select on move is enabled and when using the arrow keys. If that's not the case, something got messed up, and that'd be surprising to me.

DaniGuardiola replied:
See https://github.com/WordPress/gutenberg/pull/65387#discussion_r1797694534 😅 

Will fix soon. Silly me!

ciampo replied:
Fix in https://github.com/WordPress/gutenberg/pull/66198 — it should also fix the pattern categories in the block sidebar mentioned by @t-hamano above.

DaniGuardiola commented on packages/components/src/tabs/styles.ts:
```diff
@@ -175,23 +195,69 @@ export const Tab = styled( Ariakit.Tab )`
+		[data-select-on-move='true']
+			[role='tab']:is( [aria-selected='true'],  )
+			& {
+			transition: opacity 0.3s ease-in;
```
@t-hamano oof, I accidentally replaced the wrong transition, this is where it should have been replaced.
